cookbook/src/tx_segwit-v0.md

```diff
@@ -0,0 +1,338 @@
+#     Witness,
+# };
+#
+# const DUMMY_UTXO_AMOUNT: u64 = 20_000_000;
+# const SPEND_AMOUNT: u64 = 5_000_000;
+# const CHANGE_AMOUNT: u64 = 14_999_000; // 1000 sat fee.
+#
+# fn senders_keys<C: Signing>(secp: &Secp256k1<C>) -> (SecretKey, WPubkeyHash) {
+#     let sk = SecretKey::new(&mut rand::thread_rng());
+#     let pk = bitcoin::PublicKey::new(sk.public_key(secp));
+#     let wpkh = pk.wpubkey_hash().expect("key is compressed");
+# 
+#     (sk, wpkh)
+# }
+# 
+# 
+# fn receivers_address() -> Address {
+#     Address::from_str("bc1q7cyrfmck2ffu2ud3rn5l5a8yv6f0chkp0zpemf")
+#         .expect("a valid address")
+#         .require_network(Network::Bitcoin)
+#         .expect("valid address for mainnet")
+# }
+# 
+# fn dummy_unspent_transaction_output(wpkh: &WPubkeyHash) -> (OutPoint, TxOut) {
+#     let script_pubkey = ScriptBuf::new_v0_p2wpkh(wpkh);
+# 
+#     let out_point = OutPoint {
+#         txid: Txid::all_zeros(), // Obviously invalid.
+#         vout: 0,
+#     };
+# 
+#     let utxo = TxOut {
+#         value: DUMMY_UTXO_AMOUNT,
+#         script_pubkey,
+#     };
+# 
+#     (out_point, utxo)
+# }
+
+fn main() {
+    let secp = Secp256k1::new();
+    let (sk, wpkh) = senders_keys(&secp);
+    let address = receivers_address();
+    let (dummy_out_point, dummy_utxo) = dummy_unspent_transaction_output(&wpkh);
+
+    // The script code required to spend a p2wpkh output.
+    let script_code = dummy_utxo
+        .script_pubkey
+        .p2wpkh_script_code()
+        .expect("valid script");
+
+    // The input for the transaction we are constructing.
+    let input = TxIn {
+        previous_output: dummy_out_point,
+        script_sig: script_code.clone(),
+        sequence: Sequence::ENABLE_RBF_NO_LOCKTIME,
+        witness: Witness::default(),
+    };
+
+    // The spend output is locked to a key controlled by the receiver.
+    let spend = TxOut {
+        value: SPEND_AMOUNT,
+        script_pubkey: address.script_pubkey(),
+    };
+
+    // The change output is locked to a key controlled by us.
+    let change = TxOut {
+        value: CHANGE_AMOUNT,
+        script_pubkey: ScriptBuf::new_v0_p2wpkh(&wpkh),
+    };
+
+    // The transaction we want to sign and broadcast.
+    let unsigned_tx = Transaction {
+        version: 2,
+        lock_time: absolute::LockTime::from_height(0).expect("valid height"),
+        input: vec![input],
+        output: vec![spend, change],
+    };
+
+    // Sign the unsigned transaction.
+    let mut sighash_cache = SighashCache::new(unsigned_tx);
+    let sighash = sighash_cache
+        .segwit_signature_hash(0, &script_code, DUMMY_UTXO_AMOUNT, EcdsaSighashType::All)
+        .expect("valid sighash");
+    let msg = Message::from(sighash);
+    let sig = secp.sign_ecdsa(&msg, &sk);
+    // Update the witness stack
+    sighash_cache
+        .witness_mut(0)
+        .expect("valid witness")
+        .push_bitcoin_signature(&sig.serialize_der(), EcdsaSighashType::All);
```

Comment:
```suggestion
        .push_bitcoin_signature(&sig.serialize_der(), EcdsaSighashType::All)
        .push(sk.public_key(&secp));
```

You mean this @tcharding?

Reply:
Need to serialize the pubkey, see https://github.com/rust-bitcoin/rust-bitcoin/pull/2084/files

Reply:
Solvet in f5d3fbda03ba7b07e21f8d256a9028872eaa6122